lib/ClangImporter/ClangClassTemplateNamePrinter.cpp

```diff
@@ -23,10 +24,14 @@ struct TemplateInstantiationNamePrinter
   NameImporter *nameImporter;
   ImportNameVersion version;
 
+  ClangImporter::Implementation &importerImpl;
+
   TemplateInstantiationNamePrinter(ASTContext &swiftCtx,
                                    NameImporter *nameImporter,
-                                   ImportNameVersion version)
-      : swiftCtx(swiftCtx), nameImporter(nameImporter), version(version) {}
+                                   ImportNameVersion version,
+                                   ClangImporter::Implementation &importerImpl)
+      : swiftCtx(swiftCtx), nameImporter(nameImporter), version(version),
+        importerImpl(importerImpl) {}
```

Comment:
I think a slightly less intrusive refactoring would be extracting `ClangImporter::Implementation::recordHasReferenceSemantics` out from its parent class into a global function `swift::importer::recordHasReferenceSemantics` – from what I can tell, it never refers to `this`, so there is no good reason for it to be a member function.